### ISSUE TITLE
Errors when using OpenBLAS through NumPy on Windows 10 2004

### ISSUE THREAD

**Issue by bashtage** [xianyi/OpenBLAS#2709]
This issue only appears on Windows 10 2004 (19041). It does not appear on Windows 10 1909 (18363).

On a fresh install of NumPy from pip on a 2004 machine, e.g,

```
pip install numpy ipython
```

open `ipython` and enter

```
import numpy as np
a = np.arange(13 * 13, dtype=np.float64)
a.shape = (13, 13)
a = a % 17
va, ve = np.linalg.eig(a)
```

This produces an error:

```
 ** On entry to DGEBAL parameter number  3 had an illegal value
 ** On entry to DGEHRD  parameter number  2 had an illegal value
 ** On entry to DORGHR DORGQR parameter number  2 had an illegal value
 ** On entry to DHSEQR parameter number  4 had an illegal value
---------------------------------------------------------------------------
LinAlgError                               Traceback (most recent call last)
<ipython-input-1-bad305f0dfc7> in <module>
      3 a.shape = (13, 13)
      4 a = a % 17
----> 5 va, ve = np.linalg.eig(a)

<__array_function__ internals> in eig(*args, **kwargs)

c:\anaconda\envs\py-pip\lib\site-packages\numpy\linalg\linalg.py in eig(a)
   1322         _raise_linalgerror_eigenvalues_nonconvergence)
   1323     signature = 'D->DD' if isComplexType(t) else 'd->DD'
-> 1324     w, vt = _umath_linalg.eig(a, signature=signature, extobj=extobj)
   1325
   1326     if not isComplexType(t) and all(w.imag == 0.0):

c:\anaconda\envs\py-pip\lib\site-packages\numpy\linalg\linalg.py in _raise_linalgerror_eigenvalues_nonconvergence(err, flag)
     92
     93 def _raise_linalgerror_eigenvalues_nonconvergence(err, flag):
---> 94     raise LinAlgError("Eigenvalues did not converge")
     95
     96 def _raise_linalgerror_svd_nonconvergence(err, flag):

LinAlgError: Eigenvalues did not converge
```

What makes me suspect that it is OpenBLAS related is:

1. It does not happen when using MKL.
2. It does not happen when using LAPACK-lite that ships with NumPy and is used as a fallback.

On the other hand, it may likely be a Windows bug since:

1. It does not occur before 2004.
2. It does not occur when using 32-bit Python on the same system.
3. It does not occur when running Pythin in WSL on the same machine whether using WSL1, which runs Linux binaries directly on Windows, or WSL2, which runs them in a hypervisor.

Any help is appreciated.

The related NumPy issue is numpy/numpy#16744.

**Comment by martin-frbg** [comment]
While I agree that it looks "OpenBLAS related", I am not yet convinced it is a bug in OpenBLAS itself, particularly if it only happens with/after the Win10 update. The error messages look as if argument passing from C (BLAS) to FORTRAN (LAPACK) functions took a hit. 
Perhaps it would make sense to try and build OpenBLAS on a Win10-2004 system (assuming that your `pip install` just pulled a precompiled binary). Btw it might also be relevant if this affects only particular cpu generations.

**Comment by bashtage** [comment]
I ran openblas_utext.exe without any issues.  Is there an equivalent to the `lapack-test` directive in the *nix make file?

**Comment by martin-frbg** [comment]
The executables in `test` (if you have them) are at least fortran files calling into C routines, but the lapack-test is currently not available in the cmake build. I'll see if it is sufficient to add its subdirectory to the main CMakeLists.txt but I suspect some things may need adjusting compared to the stock cmake support in Reference-LAPACK.

**Comment by brada4** [comment]
Indeed pip will install binary by default.
It needs parameter `--no-binary :all:` to force compilation of all module libraries.

The illegal value parameters are all integers. This might mean that LAPACK and BLAS has different idea of integer sizes. Could you copy numpy directory from working version and test again?It just looks as broken wheel package pulled from pypi.

**Comment by bashtage** [comment]
I tried the nightly from 5/7/20 but this also failed.  It has a different version than 1.19 although it isn't obvious how it differs other than it has a different hash.  

I'll try to build myself in the next few days.

**Comment by brada4** [comment]
It is not directly made by OpenBLAS, could you point to hashes you talk about so that we can have a look at package content?

https://pypi.org/project/numpy/#files

**Comment by brada4** [comment]
@bashtage what would be interesting to compare numpy wheels (' hashes) in pip caches. Looking into archives they include 0.3.9-dev version which is anything from development tree (maybe known broken) between 0.3.9 and 0.3.10. 
I think we need to add git tag as once proposed out of fact that downstreams do not stick to the releases.

**Comment by bashtage** [comment]
From NumPy 1.19.0:

libopenblas.NOIJJG62EMASZI6NYURL6JBKM4EVBGM7.gfortran-win_amd64.dll

`strings` indicates this is 3.9

From NumPy nightly 5/7/2020:

libopenblas.VN4TFHCG6GCB7E53NOJHJWQ4PL7VXRV3.gfortran-win_amd64

`strings` indicates this is 3.10

[It is my assumption that the large string is a hash]

**Comment by martin-frbg** [comment]
numpy should know which hash they packaged, most likely it is a snapshot from just a few days before the 0.3.10 release to get the fix for inadvertent use of avx512 on haswell (in dynamic_arch builds created on skx). most likely a red herring when the issue seems so closely tied with a microsoft patch.

**Comment by brada4** [comment]
Is the hash on the working system matching any of these two?

**Comment by martin-frbg** [comment]
There is a (small) chance that #2729 may have fixed this (inconsistent declarations of the size of the GEMM buffer in DYNAMIC_ARCH builds)

**Comment by martin-frbg** [comment]
@bashtage Any news on this ?

**Comment by bashtage** [comment]
NumPy is using head in their latest nightlies and it is still producing the same issues.

Edit: I do think it is more and more likely a Windows bug.  It even appears when coretype is set to Prescott. I had planned to reach out to a Microsoft Python advocate to see if I can get any bandwidth beyond reporting it through standard channels. It won't show up in CI for a very long time since most CI instances are using build 17xxx of Windows, while 19041+ is required for the bug to appear.

**Comment by brada4** [comment]
> Prescott

What is your CPU? Like CPU-Z screenshot would be nice
Should be fixable setting OPENBLAS_CORETYPE= SKYLAKEX/HASWELL/SANDYBRIDGE for AVX512/AVX2/AVX respectively.

**Comment by bashtage** [comment]
I'm on Zen by default.  I see the same errors by manually setting core
types to Haswell, Sandybridge, Zen and Nehalem.

**Comment by brada4** [comment]
So it should be ZEN. there is recent change to detect 3td generation ZEN CPUs #2744 , coming downstream in due time.
The illegal value error indeed is disturbing.

**Comment by martin-frbg** [comment]
@brada4 this has all already been done and discussed here and on the numpy ticket and I do not see what Zen3 has to do with it

**Comment by mattip** [comment]
A little more information. I used the instructions in [the wiki](https://github.com/xianyi/OpenBLAS/wiki/How-to-use-OpenBLAS-in-Microsoft-Visual-Studio) to use CMake and vstudio 2019 to build OpenBLAS. The error still reproduces. 

Tracking it down, the first error is [here](https://github.com/xianyi/OpenBLAS/blob/v0.3.10/lapack-netlib/SRC/dgebal.f#L319) in a call to `DNRM2` which results in a NAN even though the data seems valid. (I used `WRITE(*,*) A` in a loop just to make sure). My fortran foo is not strong enough to figure out which `DNRM2` is being called. It is neither the one in `reference/dnrm2f.f` nor the one in `lapack-netlib/BLAS/SRC/dnrm2.f`: adding `WRITE(*,*) N` to the top of both those function did not print anything. Any hints as to how I can find where the implementation might be? Might there be, in one of the implementations of this function, a variable that is not initialized when it is compiled on windows?

**Comment by martin-frbg** [comment]
The two Fortran files you found are both from the original reference (or "netlib") BLAS (one "accidentally" included as a part of Reference-LAPACK, the other dead code from an earlier idea to provide a built-in verification). 
The code that actually gets called is interface/nrm2.c which forwards to the appropriate cpu-specific microkernel, named NRM2_K in the file (which on x86_64 with a non-Microsoft compiler would be kernel/x86_64/nrm2.S according to the KERNEL files in the x86_64 directory). As you are building with VS,
which does not cope with the assembly code, the generic implementation from kernel/arm/nrm2.c is substituted.
(the CNAME macro there gets replaced with DNRM2, and FLOAT with double). That code looks quite straightforward to me, but maybe instrumenting it with print statements will provide some insight ?

**Comment by mattip** [comment]
Surprisingly, adding print statements to `kernle/arm/nrm2.c` did not do anything. I am pretty sure the build _is_ using the assembly version via clang (maybe because of flang?), and in the numpy builds it is using mingw, so maybe also getting the assembly kernel. I see compiled kernels in How can I convince OpenBLAS to use the C version instead, as a way to check whether there is some new problem with the assembly compilation or register use, or is there a way to add a `fprint("A\n")` to the assembly to positively prove I am using it?

Is there a set of tests I can run in OpenBLAS that might expose the problem (sorry for the noise if you have already tried that)? I wonder if it is critical that the matrix be 13 x 13 ?

**Comment by martin-frbg** [comment]
With mingw or clang you can add a line
`DNRM2KERNEL=../arm/nrm2.c`
to kernel/x86_64/KERNEL.HASWELL (or whatever matches your cpu)

**Comment by mattip** [comment]
My cpu is "model name	: AMD Ryzen 9 3900X 12-Core Processor" so I assume that is KERNEL.ZEN.
What do I need to touch to get that to be rebuilt? I don't seem to be able to trigger the replacement of the assembly version with the C version.

**Comment by martin-frbg** [comment]
`make clean` should do it - if you are using cmake, it may be necessary to recreate the build directory however

**Comment by mattip** [comment]
Forcing the use of the C-based `../arm/nrm2.c` kernel fixes the dnrm-related failures. When running the entire NumPy test suite with the patch in the detail, I still have one failure: `dgelsd` reports `** On entry to DLASCL parameter number  4 had an illegal value`

<details>

```
diff --git a/kernel/x86_64/KERNEL b/kernel/x86_64/KERNEL
index 4a2e13be..e1b1e6e4 100644
--- a/kernel/x86_64/KERNEL
+++ b/kernel/x86_64/KERNEL
@@ -259,7 +259,8 @@ SNRM2KERNEL = nrm2_sse.S
 endif

 ifndef DNRM2KERNEL
-DNRM2KERNEL = nrm2.S
+//DNRM2KERNEL = nrm2.S
+DNRM2KERNEL      =  ../arm/nrm2.c
 endif

 ifndef QNRM2KERNEL

```

**Comment by martin-frbg** [comment]
That's another NaN then - the check in DLASCL is for double argument "cfrom" zero or NaN. And we still do not really know where they come from, or do we ?

**Comment by mattip** [comment]
Looking at the registers and flags in a debugger, the only thing that seems strange at the point where NumPy calls in to OpenBLAS is that ST0 is NAN when `fmod()` is used, and `+0` when it is not used.

**Comment by mattip** [comment]
It is legal to call [`fstp`](https://www.felixcloutier.com/x86/fst:fstp) without previously pushing anything onto the fpu stack? [Here](https://gist.github.com/mattip/d9e1f3f88ce77b9fde6a285d585c738e) is the first part of the implementation of `fmod`, the `ST0` register gets messed up on line 98

**Comment by bashtage** [comment]
> Looking at the registers and flags in a debugger, the only thing that seems strange at the point where NumPy calls in to OpenBLAS is that ST0 is NAN when `fmod()` is used, and `+0` when it is not used.

Any idea if this register is NaN under Linux?  Guessing no.

**Comment by carlkl** [comment]
Maybe this gives some idea:

https://stackoverflow.com/questions/48332763/where-does-the-xmm-instruction-divsd-store-the-remainder

`...
A problem you may run into in using FPREM is the fact that you can't mix code using the Floating Point Unit and the MMX registers. They both end up using the same registers, but differently.
...`
`Transition from x87 FPU to MMX technology instructions or to SSE or SSE2 instructions that operate on MMX registers should be preceded by saving the state of the x87 FPU.
...`

**Comment by carlkl** [comment]
Here are some alternative fmod implementations to compare:

https://git.musl-libc.org/cgit/musl/tree/src/math/fmod.c

https://github.com/JuliaMath/openlibm/blob/cca41bc1abd01804afa4862bbd2c79cc9803171a/src/e_fmod.c

https://github.com/JuliaMath/openlibm/blob/cca41bc1abd01804afa4862bbd2c79cc9803171a/i387/e_fmod.S

https://github.com/mirror/mingw-w64/blob/16151c441e89081fd398270bb888511ebef6fb35/mingw-w64-crt/math/x86/fmod.c

fmod taken from mingw-w64 libmingwex.a:

```ams
sub    $0x18,%rsp
movsd  %xmm0,0x8(%rsp)
fldl   0x8(%rsp)
movsd  %xmm1,0x8(%rsp)
fldl   0x8(%rsp)
fxch   %st(1)
fprem  
fstsw  %ax
sahf   
jp     1a <fmod+0x1a>
fstp   %st(1)
fstpl  0x8(%rsp)
movsd  0x8(%rsp),%xmm0
add    $0x18,%rsp
retq   
```

**Comment by mattip** [comment]
Thanks @carlkl. I think those alternative would be better on the NumPy issue numpy/numpy#16744, or even a PR to blocklist the fmod implementation and provide a substitute. NumPy does this already for problematic functions, grep around for "blocklist". 

As for OpenBLAS: I am not sure we can point only at fmod for causing this problem; if there is some register clearing or untangling that needs to be done, it should be done in OpenBLAS to shield itself from other libraries messing things up.

**Comment by brada4** [comment]
@mattip  yep, every now and then some unitialized usage is corrected. It is sort of automatic in C-only code, but assembly needs in-depth review of almost every piece.

**Comment by martin-frbg** [comment]
nrm2.S is unchanged from the GotoBLAS days. As I read it, it does not save/restore fpu state but at least if starts with putting a bunch of zeroes on the fpu stack so should not read any pre-existing garbage (?). Could try to add an fsave instruction after the preamble I guess.
I wonder what the backtrace for the DLASCL error looks like, at least from their name all the other "older" assembly kernels for x86_64 appear to be sse2 (and employ a SAVEREGISTERS macro in Windows builds
to save registers 6-15. I cannot seem to find anything about similar requirements for fpu state on Windows)

**Comment by mattip** [comment]
> I wonder what the backtrace for the `DLASCL` error looks like

The call chain is numpy calls zgelsd [calls zlalsd](https://github.com/xianyi/OpenBLAS/blob/develop/lapack-netlib/SRC/zgelsd.f#L517) which [calls dlanst](https://github.com/xianyi/OpenBLAS/blob/develop/lapack-netlib/SRC/zlalsd.f#L310) where [`D( N )` is already NAN](https://github.com/xianyi/OpenBLAS/blob/develop/lapack-netlib/SRC/dlanst.f#L144) so it returns NAN into `ORGNRM` in zlalsd.

**Comment by martin-frbg** [comment]
Thank you. Seems ZLALSD has some chance to call ZDROT prior to calling DLANST, which would be zrot_sse2.S for Zen. Could set ZROTKERNEL=../arm/zrot.c in kernel/x86_64/KERNEL.ZEN  (or more globally change its definition in KERNEL) to try.
(But if ZGELSD got NaNs already perhaps this would need to be addressed in Reference-LAPACK)

**Comment by bashtage** [comment]
Not sure if is useful, but the lapack lite that comes with NumPy and is used when OpenBLAS and MKL are not available does not experience this issue.  In fact, MKL does not either.

**Comment by brada4** [comment]
Which BLAS implementation is used to build NumPy?
```
import numpy
numpy.show_config()
```

**Comment by mattip** [comment]
> Could set ZROTKERNEL=../arm/zrot.c in kernel/x86_64/KERNEL.ZEN

Nope, it did not clear the "invalid argument" message.

**Comment by mattip** [comment]
> if ZGELSD got NaNs already  ...

This test passes on other platforms and used to pass on windows, so I don't think the problem is due to NaNs in the input.

**Comment by martin-frbg** [comment]
Strange as I see no other BLAS calls between ZLALSD and DLANST, and all LAPACK functions in between are taken directly from Reference-LAPACK (which I assume is identical to the numpy "lapack lite"). ZLASET checks its input for NaN but will generate NaNs itself if both "FROM" and "TO" arguments are infinite.

**Comment by carlkl** [comment]
What we know is that calling fmod from ucrtbase.dll 10.0.19041.488 (Windows version 2004) **definitely triggers an FPU exception** under certain circumstances. This is quoted from https://developercommunity.visualstudio.com/content/problem/1208774/fpu-exception-in-fmod0-x-in-windows-10-version-200.html _FPU exception in fmod(0., x) in Windows 10 version 2004_ 
and related to https://developercommunity.visualstudio.com/content/problem/1207405/fmod-after-an-update-to-windows-2004-is-causing-a.html _fmod(), after an update to windows 2004, is causing a strange interaction with other code_

Reading https://randomascii.wordpress.com/2016/09/16/everything-old-is-new-again-and-a-compiler-bug/ you can find the following statement:

_The x87 FPU started life as a coprocessor. Because of this (or maybe it was because of floating-point 
latencies – doesn’t matter) it doesn’t report exceptions synchronously._ 
_So, in the crazy world of x87 floating-point, exceptions are reported on the next x87 instruction that executes._
_The x87 designers weren’t totally cruel and capricious. They dedicated a register to storing the address of the 
last x87 instruction that executed – the actual problematic instruction!_

or here: http://masm32.com/board/index.php?topic=186.0 

_Bits 6-0 are flags raised by the FPU whenever it detects an exception. Those exception flags are cumulative 
in the sense that, once set (bit=1), they are not reset (bit=0) by the result of a subsequent instruction which, 
by itself, would not have raised that flag. Those flags can only be reset by either initializing the FPU (FINIT instruction) 
or by explicitly clearing those flags (FCLEX instruction)._ 

_Which means that if you check for an invalid operation at the end of a series of FPU instructions, that flag could 
have been raised by any of those previous instructions and not necessarily by the last one._ 

So - long story - could it be possible that a subsequent FPU instruction somewhere in the OpenBLAS codebase
segfault due to the fmod exception?

**Comment by martin-frbg** [comment]
I had already offered adding fninit at the start of the OpenBLAS nrm2.S on the numpy ticket, but (understandably) everybody there seems to be fixated on solving this on the numpy side. :) 

I do not think any of the other BLAS microkernels for ZEN uses x87 fpu instructions, thus the DLASCL
problem is still unclear (unless gfortran decided to use fpu in compilation of one of the LAPACK files, 
or SSE2 checks the same old exception register. If I recall the discussion above correctly, that problem also survived replacement of ucrtbase.dll)

**Comment by bashtage** [comment]
> I do not think any of the other BLAS microkernels for ZEN uses x87 fpu instructions, thus the DLASCL
> problem is still unclear (unless gfortran decided to use fpu in compilation of one of the LAPACK files,
> or SSE2 checks the same old exception register. If I recall the discussion above correctly, that problem also survived replacement of ucrtbase.dll)

Not sure if this is relevant, but I think this bug hits all cores except _very_ old ones (Nehalem??, sorry don't remember).

**Comment by martin-frbg** [comment]
Thanks, probably not relevant (ZEN just being what the current testbed is IIRC), OTOH I understand the
DLASCL bug is _also_ something that only happens with Windows10-2004 - or am I wrong ?

**Comment by bashtage** [comment]
@martin-frbg Yes, that is correct.  Only the AMD64 variant on Windows 10 2004, and then seemingly only after fmod is called.

**Comment by bashtage** [comment]
I was wondering if there is something common in the more modern kernels that is missing in the older ones that might give a hit as to the source of the bug.  Also curious why MKL seems to work OK, but that can't be answered here.

**Comment by martin-frbg** [comment]
They all use the same code for NRM2, and there is no other BLAS call except the ZROT (that was already shown to be irrelevant) on the code path leading to DLASCL complaining.

**Comment by mattip** [comment]
Putting a `fninit` into `nrm2.S` fixes that part.

<details>
```
diff --git a/kernel/x86_64/nrm2.S b/kernel/x86_64/nrm2.S
index e9be1262..548e3b74 100644
--- a/kernel/x86_64/nrm2.S
+++ b/kernel/x86_64/nrm2.S
@@ -50,6 +50,7 @@
        PROLOGUE
        PROFCODE

+       fninit
        fldz
        testq   M, M
        jle     .L999
```

**Comment by mattip** [comment]
There are 27 files in kernel/x86_64 that use `fldz`. I will try putting `fninit` into all of them, what is there to lose.

**Comment by mattip** [comment]
> fixes that part.

... but there is still one test error in `zlalsd` (sorry I left that part out).

**Comment by martin-frbg** [comment]
Thx, so it is indeed "fixable" in the callee. The FNCLEX instruction that carlkl dug up should be preferable though (far fewer ops according to Agner Fog's tables) - and the other .S files with fpu instructions in them should be unused on ZEN at least , so probably irrelevant to the case.
The ZLALSD test error is the same "DLASCL raising an invalid argument # 4 error" we had above ?

**Comment by mattip** [comment]
yes

**Comment by mattip** [comment]
If `fninit` works cleanly we can try to  optimize with `fnclex`

**Comment by carlkl** [comment]
Great to hear. All this seems to be a fundamental Microsoft problem. It must not be that a VC runtime library spams the FPU registers/flags. And this is not the first time as far I can remember.

**Comment by mattip** [comment]
Putting `fninit` at the top of the ~27 files makes numpy tests pass cleanly. `fnclex` does not clear any of the failing tests.

**Comment by martin-frbg** [comment]
Interesting, thanks, So who was the hidden call between ZLALSD and DLASCL if it wasn't zrot ?
(Pity that fnclex did not work - according to Agner Fog's Instruction Tables for x86_64, fninit is about 4 times more expensive)
May still want to make that ifdef OS_WINDOWS or something (OTOH I dimly remember another issue where one calculation would magically inject NaNs into another, and that was on Linux - think xianyi just disabled one of the "offending" assembly kernels back then, so may be worth a revisit)

**Comment by carlkl** [comment]
This is really great news. I have some concerns, that leaving the FPU switched to extended precision may pop up other issues - but maybe not in WIN64 - who knows? It is to late today for me to find an alternative solution.
EDIT: Maybe fnclex  plus clearing the the FPU stack?

**Comment by mattip** [comment]
Another solution may be to use C code instead of these kernels. Are we sure the assembly is still faster (it may have been at one time, but some of these are quite old)?

**Comment by martin-frbg** [comment]
(#695 is what I had somewhere in the back of my mind, but at first glance it does/did not involve any fpu-using microkernels).
Yes, using the generic c codes on Windows would probably be easier (need to check if we can ifdef on OSNAME in the KERNEL files already)

**Comment by carlkl** [comment]
I know I am persistent, but there is a difference between `FCLEX`  and `FNCLEX`:

```
Clears the floating-point exception flags (PE, UE, OE, ZE, DE, and IE), the exception summary status flag (ES), 
the stack fault flag (SF), and the busy flag (B) in the FPU status word. The FCLEX instruction checks for and 
handles any pending unmasked floating-point exceptions before clearing the exception flags; 
the FNCLEX instruction does not.

The assembler issues two instructions for the FCLEX instruction (an FWAIT instruction followed by an FNCLEX instruction), 
and the processor executes each of these instructions separately. If an exception is generated for either of these 
instructions, the save EIP points to the instruction that caused the exception.
```

Maybe one should `FCLEX` give a try.

**Comment by martin-frbg** [comment]
Maybe, but my current reading of that description is that the implied `FWAIT` in `FCLEX` would only raise the exception that we are trying to clear (and all this assumes that it is only a pending exception flag we need to worry about, not an invalid fpu stack in general)

**Comment by mattip** [comment]
> Maybe one should FCLEX give a try.

I did. It didn't solve the problem.There is a comment to that effect above.

Edit: ahh, you mean `fnclex` -> `fclex`. Sorry.

**Comment by martin-frbg** [comment]
Seems most likely to me now that the mystery .S file "causing" the ZLALSD/DLASCL test failure is simply znrm2.S , all the other files with fpu instructions are either completely dead or extensions unlikely to be called from numpy

**Comment by carlkl** [comment]
I have taken the trouble to disassemble all object files from https://github.com/xianyi/OpenBLAS/releases/download/v0.3.10/OpenBLAS-0.3.10-x64.zip and take a look at FPU and MMX usage:

```
FPU instructions:

cblas_drotg.obj
cblas_srotg.obj

crotg.obj
drotg.obj
srotg.obj
zrotg.obj

csum_k_<TARGETNAME>.obj
dsum_k_<TARGETNAME>.obj
ssum_k_<TARGETNAME>.obj
zsum_k_<TARGETNAME>.obj

dnrm2_k_<TARGETNAME>.obj
znrm2_k_<TARGETNAME>.obj

MMX instructions:

dgemm_oncopy_BARCELONA.obj.asm
dgemm_otcopy_BARCELONA.obj.asm
sgemm_oncopy_BARCELONA.obj.asm
sgemm_otcopy_BARCELONA.obj.asm
strsm_kernel_LN_BARCELONA.obj.asm
strsm_kernel_LN_CORE2.obj.asm
strsm_kernel_LN_PRESCOTT.obj.asm
strsm_kernel_LT_BARCELONA.obj.asm
strsm_kernel_LT_CORE2.obj.asm
strsm_kernel_LT_PRESCOTT.obj.asm
strsm_kernel_RN_BARCELONA.obj.asm
strsm_kernel_RN_CORE2.obj.asm
strsm_kernel_RN_PRESCOTT.obj.asm
strsm_kernel_RT_BARCELONA.obj.asm
strsm_kernel_RT_CORE2.obj.asm
strsm_kernel_RT_PRESCOTT.obj.asm
```

The following objects are using the `EMMS` instruction to **RESET** and emtpy the FPU registers after **MMX** usage:

```
dgemm_oncopy_BARCELONA.obj.asm
dgemm_otcopy_BARCELONA.obj.asm
sgemm_oncopy_BARCELONA.obj.asm
sgemm_otcopy_BARCELONA.obj.asm
```
The `EMMS` instruction markes the FPU stack as emtpy and prepares it for FPU usages after use of `MMX` instructions. This is necessary as the MMX and the FPU registers are aligned. `EMMS` should also be able to prepare the FPU stack for further usage in case it is corrupted.

Probably using the `FCLEX` and `EMMS` instruction instead of `FNINIT`  in the beginning of the *nrm2.S assembler files is sufficient to heal bad behaviour from any third party library (or MS VC runtime DLL) to the FPU registers and stack. 
The benefit of these two instructions vs `FNINIT` is performance and the fact it does not change the FPU precision and is therefore consistent with `CONSISTENT_FPCSR`. This means it can be added without any considerations. 
However, this should be tested by @mattip first.

**Comment by martin-frbg** [comment]
Thanks - the xROTG ones derive from C files (in OpenBLAS/interface) so any fpu usage there is/was up to the compiler. xNRM2 and xSUM are quite familiar names now (where the xSUM is a BLAS extension probably not used by numpy). (Still more inclined to replace nrm2.S with its generic C version specifically on Windows though)

**Comment by carlkl** [comment]
Unfortunately the scattered FPU state has to be healed somewhere: in the numpy or the OpenBLAS codebase. Adding two instructions to 2 assembler files seems the most simple method to me. In numpy one has to extend the build system to add MASM, add a helper function at the right place and so on.

MS has a problematic perspective for the WIN64 API and FPU uasge: the user has to be aware of the fact tat the FPU state is volatile between calls (this includes a corrupt state I guess). So the user is responsible to get the FPU in a usuable state before usage. I'm sure nobody expected that an WIN64! On the other hand: nobody expected the spanish inquistion...

The most problematic aspect is, that the erratic FPU state is pending if not healed somewhere and may pop up later in the codebase.

**Comment by martin-frbg** [comment]
Yuck. I did not envisage OpenBLAS as a Windows repair tool but it is hard to argue about the "somewhere". Still I consider the corrupt fpu state a genuine bug in Win10.19041 and would hope Microsoft treats it as such. (And if EMMS/FCLEX actually does the trick I would still prefer to make that addition to the (z)nrm2.S files `#ifdef OS_WINDOWS`)

**Comment by brada4** [comment]
MSVC 2019 does not emit MMX or FPU in 64bit mode.

**Comment by martin-frbg** [comment]
Does not seem to make it safer to just leave the pending exception dangling, at best it would make it somebody else's problem ? (Assuming some other third-pary library gets loaded as well, like we saw with libhdfs in another recent issue)

**Comment by carlkl** [comment]
@brada, yes, MSVC does not emit MMX or FPU in 64bit mode. But the MS runtime libraries are using FPU code and the use it in an unexpected way, hence the problems. As long as OpenBLAS is using FPU or MMX code itself in Win64 in the gcc builds OpenBLAS must care about this problem.
FCLEX  and EMMS instructions are not harmful and costs only some cycles. So why not add them in the right place?

**Comment by martin-frbg** [comment]
Do we know by know that the FCLEX (or FNCLEX) plut EMMS combo is sufficient to clear this sorry state ?

**Comment by carlkl** [comment]
I will try it as soon as I can. I have to setup my development machine first.

**Comment by martin-frbg** [comment]
Further to this, do we know that routines using only MMX are definitely affected by this DLL bug as well, or is this just an assumption based on them sharing the same registers ? (Thinking MMX is a bit more likely to still be in use than x87, so there should be more users and programs affected by the bug)

**Comment by carlkl** [comment]
I will try it as soon as I can. I have to setup my development machine first.

**Comment by carlkl** [comment]
@martin-frbg, I'm not a assembler guru. My idee what happens with this crazy issue are described here: https://github.com/xianyi/OpenBLAS/issues/2709#issuecomment-703675777 and are the results of reflection after scamming numerous documents, stackoverflow/blog contributions. A huge time sink btw. **EDIT**: we should file MS an invoice :-(

I have an idea about why MMX is not affected: Maybe because MMX uses the registers as is and not as a circulating stack. But this is speculation. So from the MMX point of view the registers can be simply used and erroneous values will be overwritten. 

The documentation says:

_The EMMS instruction must be used to clear the MMX technology state at the end of all MMX technology procedures or subroutines and before calling other procedures or subroutines that may execute x87 floating-point instructions. If a floating-point instruction loads one of the registers in the x87 FPU data register stack before the x87 FPU tag word has been reset by the EMMS instruction, an x87 floating-point register stack overflow can occur that will result in an x87 floating-point exception or incorrect result._

_EMMS operation is the same in non-64-bit modes and 64-bit mode._

I have drawn two conclusions from that:

1.  Before using the FPU EMMS can heal the the FPU state.
2.  This instruction is not needed for MMX usage

But again: I'm not an assembler guru.

**Comment by brada4** [comment]
> ---, yes, MSVC does not emit MMX or FPU in 64bit mode. But the MS runtime libraries are using FPU code and the use it in an unexpected way, hence the problems. As long as OpenBLAS is using FPU or MMX code itself in Win64 in the gcc builds OpenBLAS must care about this problem.
> FCLEX and EMMS instructions are not harmful and costs only some cycles. So why not add them in the right place?

It looks like somehow the stuf that applied to userspace WDM (printer drivers) that mmx state is corrupted over context switch, came to generic applications too. Would be interesting to see if windows vista compatibility layer fixes this.

**Comment by carlkl** [comment]
@brada, the error has been localized see https://github.com/numpy/numpy/issues/16744#issuecomment-702053207. It is caused by the use of fmod from ucrtbase.dll (10.0.19041.488). See https://developercommunity.visualstudio.com/content/problem/1207405/fmod-after-an-update-to-windows-2004-is-causing-a.html and https://developercommunity.visualstudio.com/content/problem/1208774/fpu-exception-in-fmod0-x-in-windows-10-version-200.html.

**Comment by martin-frbg** [comment]
@carlkl think you may be tagging an unrelated user in your replies to brada4 here. @brada4 from your suggestions I am not sure if you have been following this thread...

**Comment by carlkl** [comment]
@martin-frbg, I think you are right.

**Comment by mattip** [comment]
Commenting on this closed issue to update that in numpy/numpy#17547 I added code to call `emms` after a call to fmod in windows, so now we have both a belt and suspenders (does that work in English?)

**Comment by bashtage** [comment]
@mattip <uk translation>Belt and bracers</uk translation>
